rollup/ccc/async_checker.go

```diff
@@ -0,0 +1,175 @@
+package ccc
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sourcegraph/conc/stream"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core"
+	"github.com/scroll-tech/go-ethereum/core/rawdb"
+	"github.com/scroll-tech/go-ethereum/core/state"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/core/vm"
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/metrics"
+	"github.com/scroll-tech/go-ethereum/params"
+	"github.com/scroll-tech/go-ethereum/rollup/tracing"
+)
+
+var (
+	failCounter        = metrics.NewRegisteredCounter("ccc/async/fail", nil)
+	checkTimer         = metrics.NewRegisteredTimer("ccc/async/check", nil)
+	activeWorkersGauge = metrics.NewRegisteredGauge("ccc/async/active_workers", nil)
+)
+
+type Blockchain interface {
+	Database() ethdb.Database
+	GetBlock(hash common.Hash, number uint64) *types.Block
+	StateAt(root common.Hash) (*state.StateDB, error)
+	Config() *params.ChainConfig
+	GetVMConfig() *vm.Config
+	core.ChainContext
+}
+
+// AsyncChecker allows a caller to spawn CCC verification tasks
+type AsyncChecker struct {
+	bc             Blockchain
+	onFailingBlock func(*types.Block, error)
+
+	workers      *stream.Stream
+	freeCheckers chan *Checker
+}
+
+type ErrorWithTxnIdx struct {
```

Comment:
How will we handle this?

My concern is: If this happens, shouldn't we terminate all running tasks, empty the async checker pipeline, and resubmit the new tasks once the blocks are updated?

Reply:
> shouldn't we terminate all running tasks, empty the async checker pipeline, and resubmit the new tasks once the blocks are updated?

ideally yes, the way that it is implemented here is a bit wasteful. But I want to keep first iteration as simple as possible.

Currently, worker just ignores a CCC failure if the failing block is not canonical. https://github.com/scroll-tech/go-ethereum/blob/d2eb292fd21eea5a9b4548bb6daaea3999397e00/miner/scroll_worker.go#L937-L946

> How will we handle this?

here is an example. https://github.com/scroll-tech/go-ethereum/blob/d2eb292fd21eea5a9b4548bb6daaea3999397e00/miner/scroll_worker.go#L584-L596